File: leetcode-cn/25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#


# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        new_head = ListNode(0)
        new_head.next = head

        def swap(head, tail):
            pre = tail.next
            p = head
            while pre != tail:
                next = p.next

                p.next = pre
                pre = p

                p = next

            return tail, head

        pre = new_head
        tail = pre

        cnt = 0
        while tail:
            cnt += 1
            tail = tail.next

            if not tail:
                return new_head.next

            if cnt == k:
                next = tail.next
                head, tail = swap(head, tail)

                pre.next = head
                tail.next = next

                pre = tail
                head = next

                cnt = 0

        return new_head.next


# @lc code=end

File: leetcode-cn/C++/25.k-个一组翻转链表.cpp
/*
 * @lc app=leetcode.cn id=25 lang=cpp
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
#include <vector>
using namespace std;

// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode()
//         : val(0), next(nullptr)
//     {
//     }
//     ListNode(int x)
//         : val(x), next(nullptr)
//     {
//     }
//     ListNode(int x, ListNode *next)
//         : val(x), next(next)
//     {
//     }
// };

class Solution
{
public:
    pair<ListNode *, ListNode *> reverse(ListNode *head, ListNode *tail)
    {
        ListNode *pre = tail->next;
        ListNode *cur = head;

        while (pre != tail)
        {
            ListNode *next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }

        return {tail, head};
    }

    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode *new_head = new ListNode();
        new_head->next = head;

        ListNode *pre = new_head;
        ListNode *tail = new_head;

        while (1)
        {
            for (int i = 0; i < k; i++)
            {
                tail = tail->next;
                if (tail == nullptr)
                {
                    return new_head->next;
                }
            }

            pair<ListNode *, ListNode *> ret = reverse(head, tail);
            head = ret.first;
            tail = ret.second;

            pre->next = head;
            pre = tail;
            head = tail->next;
        }

        return new_head->next;
    }
};
// @lc code=end
